pkg/ddl/id_allocator.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2013 The ql Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSES/QL-LICENSE file.
+
+package ddl
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/kv"
+	"github.com/pingcap/tidb/pkg/meta"
+)
+
+var (
+	// IDAllocatorStep is the default step for id allocator.
+	IDAllocatorStep = 100
+)
+
+// IDAllocator is used to allocate global ID.
+type IDAllocator interface {
+	AllocIDs(n int) ([]int64, error)
+}
+
+type allocator struct {
+	cache []int64
```

Comment:
If the cache exists, if there are multiple tidbs in a cluster, is the following possible?
TiDB1: Execute DDL1, DDL2...DDL201 on table tblx. ,
TiDB2: Executes DDL302 (job.ID: 302) on table tbl;
After 5 minutes, DDL201 is not finished
TiDB1: Execute DDLn (job.ID: 202) on table tbl. ,

Then DDLn will start executing 5 minutes later than DDL302, but the actual execution will start and finish first.
There's no error problem, but it's weird. In addition, this ID will also be used for the ID of the table, which may also have similar problems.

Reply:
Yes, that's possible. See id_allocator_test.go. If multiple tidbs execute ddl at the same time, the order of job ids is no longer guaranteed. And there may be id gaps between two jobs

Reply:
However, the previous implementation also had similar problems. For example, the job that requested the ID later was executed for a long time before being committed. The order of the IDs was not guaranteed and had no impact.

Reply:
The original ID order is not guaranteed, but the interval is relatively small. It doesn't seem to be a problem at the moment, but this feels kind of weird. table ID also has this problem. Does this give an optimized performance comparison? See if the optimization effect is obvious.

Reply:
Performance is one aspect, but more is robustness. When multiple tidbs allocate IDs at the same time, it will cause a lot of write conflicts, and changing to pessimistic transactions will also affect performance.  see https://github.com/pingcap/tidb/issues/27197

Reply:
1. If you change step to 10, maybe the cases I mentioned above are less.
2. The occurrence probability of this scenario(#27197) you mentioned is relatively small. In addition, theoretically we have limitDDLJobs, so this kind of problem should not occur for a TiDB. But we use the same key for tableID and jobID, so a create table operation may result in multiple ID assignments. 
Suggestion: If we put tableID and jobID assignments together, this conflict will be reduced a lot. And, it does not affect the old logic (which considers the ID assignment to be relatively incremental).

Reply:
In fact, we already have global id lock in the original logic, which means that there will no longer be a write conflict in the same tidb, but in multiple tidbs, this problem cannot be avoided.

Reply:
`txn.SetOption(kv.Pessimistic, true)` doesn't mean pessimistic transactions, we have to lock the required key explicitly, i tested `lock + incr(global id)` using 1000 routines, qps is about 1k

> Suggestion: If we put tableID and jobID assignments together, this conflict will be reduced a lot

agree with this, we can consider caching id after this optimization, if this part is still worth optimize.

FYI i have a on-going pr separating local ddl jobs from normal ones in `addBatchDDLJobs`, and `change job id allocation + insert into table` run in one txn, to make sure job submitted in id order